[needs-docs] Ensure layer legend graphics are horizontally scrollable (fixes #28050)

dakcarto opened the pull request:
## Description
This fixes #28050 by allowing the Layers panel QTreeView to horizontally scroll.

![layer-panel_horiz-scroll2](https://user-images.githubusercontent.com/1295919/59541490-c1ed5e00-8ebe-11e9-87fc-91d7d316a870.gif)

It also offers the following:
- Layer titles can be fully viewed (no elide).
- The width of the layer panel no longer needs resized via its splitter as often, reducing canvas redraws.
- When the tree view has many groups or nesting, it is simpler to scroll right than resize panel.

Even though the changes are small, the result may be confusing for some users, because users can end up with the following, even though layers are loaded:

<img width="185" alt="layers-panel_scrolled-blank" src="https://user-images.githubusercontent.com/1295919/59541529-f06b3900-8ebe-11e9-9d7f-cb475c702122.png">

To solve this, I tinkered with the following:
- Auto-reset horizontal scrollbar to position 0 after a delay, but not while user is still scrolling or has the input device pressed. While the reset worked, the timer could trigger at inopportune times.
- Add a scroll area to _just_ the WMS graphic. This proved to be tricky and not as simple/clean as scrolling the whole tree.

**Proposed horizontal scrollbar reset solution**
Assuming that just leaving it as is (user in control of scrolling) is not acceptable UX, a small button in toolbar or tall/thin button to the left of tree view could offer a quick means of reseting the horizontal scroll bar on one click.

However, I have left a reset button out of this PR, so it can be backported to `release-3_4` branch as an initial fix.

---

The other commit, to limit the max width of embedded legend widgets could also be tied to a user setting, though this would be a 'feature'.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

Commit:
Ensure layer legend graphics are horizontally scrollable (fixes #28050)

Redmine: https://issues.qgis.org/issues/20229

Commit:
Limit the maximum width of legend embedded widgets

Since column is resized to contents, limit the expanded width of
embedded widgets, if they are not already limited.

Else, embedded widget may grow very wide due to large legend graphics.

dakcarto commented on src/gui/layertree/qgslayertreeview.cpp:
```diff
@@ -43,6 +44,10 @@ QgsLayerTreeView::QgsLayerTreeView( QWidget *parent )
 
+  // Ensure legend graphics are scrollable
+  header()->setStretchLastSection( false );
+  header()->setSectionResizeMode( QHeaderView::ResizeToContents );
```
Currently, the tree view column count is 1. The other [setSectionResizeMode call](https://doc.qt.io/qt-5/qheaderview.html#setSectionResizeMode-1), specifying column, can be used, if needed.

nyalldawson commented on src/gui/layertree/qgslayertreeview.cpp:
```diff
@@ -154,7 +159,16 @@ void QgsLayerTreeView::modelRowsInserted( const QModelIndex &index, int start, i
-          setIndexWidget( index, provider->createWidget( layer, i ) );
+          QWidget *wdgt = provider->createWidget( layer, i );
+          // Since column is resized to contents, limit the expanded width of embedded widgets,
+          // if they are not already limited, e.g. have the defaut MAX value.
```
```suggestion
          // if they are not already limited, e.g. have the default MAX value.
```

dakcarto replied:
Oops, thanks. The spell check part of prepare commit script is failing for me on macOS.

nyalldawson commented on src/gui/layertree/qgslayertreeview.cpp:
```diff
@@ -154,7 +159,16 @@ void QgsLayerTreeView::modelRowsInserted( const QModelIndex &index, int start, i
+          // TODO: Max width could be a configured setting.
+          if ( wdgt->maximumWidth() == QWIDGETSIZE_MAX )
+          {
+            wdgt->setMaximumWidth( 250 );
```
This won't scale nicely on hidpi -- can you use a font metric based max instead? (and remember to scale by Qgis::UI_SCALE_FACTOR

dakcarto replied:
I looked into that, but, for example, the Opacity slider is only part textual. How would I scale relative to hidpi for the associated widget as well?

Also, I looked into having the widget relative to the treeview viewport, but that was not pleasant at all.

dakcarto replied:
Could I just use the same text scale factor for the max width of the widget?

nyalldawson replied:
What I usually do is work out how many 'X' characters fit in the desired 250 px size on low dpi, then use 

   QFontMetrics( font() ).width( 'X') * 20 * Qgis::UI_SCALING_FACTOR

(If the desired size is about 20 characters wide, etc)

nyalldawson commented:
Trying to think outside the box -- I wonder if there's some other visual clues we could add to the scale bars here to indicate to users that there's content, it's just out of view.

Brainstorming: E.g. maybe we could utilise something along these lines:

![image](https://user-images.githubusercontent.com/1829991/59542341-a7f85d80-8f48-11e9-8a44-3bd43c3d4c27.png)

nyalldawson commented:
Also - another related change I think we should make which would assist here would be to tweak the indentation size in the tree view (also in browser), and reduce the indentation for child items. It's quite wide currently, which does encourage very horizontally wide trees..

nirvn commented:
Hmm, what impact does this have on our ever-increasingly useful indicators? That'd seem to cause a significant UX regression here.

nirvn commented:
We'd need to find a way to make indicators float above layer names to always be visible and glued to right edge of the layer tree visible area.

Commit:
Fix vertical clipping of legend graphics; scroll per pixel

Commit:
Set embedded opacity widget min/max size; add expanding spacer

Expandable subwidgets should have min/max sizes in a layout and the
appended expanding QSpacerItem ensures full width is always created in
the QTreeView item's column. In turn, this ensures layer indicators are
always over an active part of the selected item, and expandable control
subwidgets do not expand excessively.

Commit:
Draw layer tree layer item indicators floating on right of viewport

Also, remove adjustment of layer text rectangle, since veiwport scrolls
horizontally now.

Commit:
Reset minimum section size to layer tree veiwport width on resize

Since last column is resized to content (instead of stretched), the
active selection rectangle ends at width of widest visible item in tree,
regardless of which item is selected. This causes layer indicators to
become 'inactive' (not clickable and no tool tip) unless their rectangle
enters the view item's selection (active) rectangle. Always resetting
the minimum section size relative to the viewport ensures the view
item's selection rectangle extends to the right edge of the viewport,
which allows indicators to become active again.

Commit:
Draw layer tree layer indicator background, for when floating over text

Commit:
Add layer tree layer contextual menu mark, floating on right of viewport

This accomplishes several things:
- Offers an indication to user that there are layer-related actions
- Single click access to contextual menu
- Offers an indication to user where layers are located when viewport is
  horizontally scrolled to right, and layer names are not visible

dakcarto commented:
@nyalldawson, @nirvn, all suggestions implemented (excepting smaller indention), plus several other fixes. Now looks/acts like this:

![layer-panel_horiz-scroll_w-indicators](https://user-images.githubusercontent.com/1295919/59782921-3a1ba100-927c-11e9-8882-5b3670e3eb95.gif)

Please review the comments in the commits for insight on changes. The latest commit adds the layer contextual menu mark, which is accomplishes several things. It is kind of a new feature and maybe can not be backported to `release-3_4`. The other commits, IMO, can be.

The 'snap back scroll' to the left is apparently _part_ of the QTreeView implementation. In other words, anytime the user scrolls far enough to the right to no longer see the layer names, a simple click within the tree view snaps the view is back to the left. This also happens _during_ other GUI interactions, like when popping up the contextual menu.

I could not find an easy way to add highlights or cursor overrides when hovering over the indicators on the right, but did make their appearance more button-like.

The overall affect in Night Mapping UI theme is OK, but could be better. Not sure how to modify the `style.qss` to work with the indicators, since they are not objects and merely painted.

nyalldawson commented:
@dakcarto nice improvements -- I especially like the move to scroll per pixel (that's bothered me in the past too, but I didn't realise the solution was 1 line!). I'm pretty sure I've seen this same issue elsewhere in QGIS, if I can just remember where...

I'm -1 on the colored backgrounds for indicators and the : menu button. It's adding more visual noise to the legend widget, and I don't personally think it's needed.

dakcarto commented:
@nyalldawson wrote:
> I'm -1 on the colored backgrounds for indicators and the : menu button. It's adding more visual noise to the legend widget, and I don't personally think it's needed.

The indicator icons, without the background, are sometimes hard to see if they are floating over item text. I definitely prefer the background button look over a transparent background.

I see your point about the : menu button, but also think it might be useful for new users, who may not know to right-click for the contextual menu. Also, though less prevalent nowadays, some people still have a single button mouse, e.g. older Macs, which means they need to also hold a meta key modifier to get a context menu (OK, as an argument, that's a bit of stretch, though 🙂 ).

When looking into the "lines in a scroll bar" approach, I noticed that the vertical scroll bar is shown as needed. Turning it to always on will eat up space, when not needed, just to show the layer indicator lines; so, I chose to extend the indicator pattern and make the layer mark. Once I made it, it seemed a bit silly to not make it also have an action, for which I thought showing the contextual menu made the most sense.

The other approach was to simply draw a thin vertical line instead for layer node items, possibly with a thinner one representing legends and embedded widget items (all non-layer node stuff). Could get colorful, too, and use a simple color palette to indicate different node types. Then, maybe allowing the line to be a : button (for layer nodes) could still be optional in settings.

@nirvn any opinions/suggestions?

nirvn commented:
@dakcarto , nice improvements, thanks for being responsive to issues raised here. 

On the indicator background, I'm not sure there's any alternative to your approach here. I'm fine with it. Coincidentally, for a while now, the code that sets the padding for indicators to live in is broken and indicators are overlapping texts:
![Screenshot from 2019-06-20 08-39-03](https://user-images.githubusercontent.com/1728657/59812098-f639a200-9336-11e9-8519-82f076e2eb40.png)
_arggg, that is not nice_

On the vertical 3-dot indicator for each layer, I'm also -1 for a few reasons: a/ it's an already busy area, but maybe more importantly b/ it somehow defeats part of the indicators' intent, which is to be sparse enough to notice the presence of indicators to begin with. 

If you think there's still an argument to offer a shortcut for the right-click menu (touch screen users?), you could add _one_ tool button in the layer tree panel's toolbar that'd open the context menu for the active layer.

nirvn commented:
As for adding some visual cue on the scrollbar, I like it and IMHO be worth implementing to deal with the UX regression of being left with a visually blank area when the horizontal scrollbar is scrolled all the way to the right-end.

Commit:
Make layer tree layer mark just a thin line; remove context menu action

Commit:
Add padding so indicator icon does not look misaligned to background

dakcarto commented:
I removed the 3-dot button and contextual menu action and made the mark a simple thin line on the right:

![layer-panel_horiz-scroll_w-indicators-mark](https://user-images.githubusercontent.com/1295919/59867133-9a791400-934a-11e9-8501-65113c3a28da.gif)

Moving the layer mark(s) into the vertical scroll bar will require it to be shown most of the time, and will take considerable more work than what I have done here. I'm OK with it as-is now, and feel all commits can be backported to `release-3_4`, but maybe needs some docs (though it seems intuitive-enough now).

NOTE: my screen recording software saves all screen objects into separate layers, so they can be manipulated independently. It seems to be mis-registering the composited cursor a bit too far to the left in the animated GIFs.

Also, added a bit of padding to icon drawing over background so the indicators look better (open in new tab and zoom in):

<img width="273" alt="layer-panel_horiz-scroll_w-indicators-mark3" src="https://user-images.githubusercontent.com/1295919/59866843-d1025f00-9349-11e9-9e9d-d3f209ca4a8b.png">

dakcarto commented:
@nyalldawson, @nirvn any followup comments on the latest iteration?

nyalldawson commented:
I'm happy for it to be merged for wider testing. We're early in the cycle anyway.

You've tagged this as a 3.4 backport -- but I don't think we should. It's quite a substantial UX change. Do you still plan on backporting this full PR?

dakcarto commented:
@nyalldawson wrote:
> I'm happy for it to be merged for wider testing. We're early in the cycle anyway.

Cool. Yeah, even though it is a 'fix,' I didn't want to try and push it into 3.8. Even though it is a small change, it definitely needs more testing and opinions. I can't think of another good means of solving the issue, and I have tried a bunch of alternatives, so far.

> You've tagged this as a 3.4 backport -- but I don't think we should. It's quite a substantial UX change. Do you still plan on backporting this full PR?

Hmm, it is substantial enough to warrant a new paragraph and some screen shots in the docs. I guess if something needs that much, it's probably a big enough UX change and shouldn't be backported.

On the flip side, it _does_ fix both horizontal and vertical clipping/scrolling issues, which would be nice to have in the LTR. So... I feel strongly both ways. 😃Since the LTR is released on a quicker cadence, I suggest merging to `master`, then seeing if the UX is (or can be) tweaked some more before considering whether it should be backported.

nirvn commented:
+1 to merge, nice work.

dakcarto commented on src/gui/layertree/qgslayertreeview.cpp:
```diff
@@ -43,13 +44,22 @@ QgsLayerTreeView::QgsLayerTreeView( QWidget *parent )
+  header()->setSectionResizeMode( QHeaderView::ResizeToContents );
+
+  // If vertically scrolling by item, legend graphics can get clipped
+  setVerticalScrollMode( QAbstractItemView::ScrollPerPixel );
```
@nyalldawson, I'm thinking that if nothing else gets backported to `release-3_4`, this should, to fix vertical scrolling.

nyalldawson replied:
@dakcarto I'm happy with that! (Also 3.8)

nyalldawson commented:
@dakcarto I noticed a regression since this was merged -- try rearranging layers by dragging them in the tree. The drop position indicator no longer works correctly.

dakcarto commented:
@nyalldawson wrote:
> I noticed a regression since this was merged -- try rearranging layers by dragging them in the tree. The drop position indicator no longer works correctly.

Hmm. I will look into it. Do you have a screen capture of it? Or, is it obvious upon drag?

nirvn commented:
@dakcarto , it'll be obvious upon drag. That black horizontal line indicating where in the tree an item being dragged would drop is simply gone.

Also, if you drag over a group, the group title area becomes a gray filled area, title gone.

dakcarto commented:
@nirvn, both issues confirmed here on macOS.

dakcarto commented:
@nyalldawson, @nirvn, fixed with https://github.com/qgis/QGIS/commit/ff77edca4494861fa545fbb6cedc3b01cf048f55
